### PR TITLE
feat(outputs): Add inlong output plugin

### LINKED ISSUE #16210: add inlong output plugin
### Use Case

Add output plugin for [Apache InLong](https://github.com/apache/inlong)

### Expected behavior

Out put to Apache InLong

### Actual behavior

Not yet supported

### Additional info

_No response_

### PR DESCRIPTION
**Pull request by justinwwhuang**
## Summary
Add [inlong](https://inlong.apache.org/) output plugin
Apache InLong is a one-stop, full-scenario integration framework for massive data that supports Data Ingestion, Data Synchronization and Data Subscription, and it provides automatic, secure and reliable data transmission capabilities. InLong also supports both batch and stream data processing at the same time, which offers great power to build data analysis, modeling and other real-time applications based on streaming data.

## Checklist
- [x] No AI generated code was used in this PR

## Related issues
resolves #16210

### CODE REVIEW

**Comment by justinwwhuang** [comment]
> Thanks so much for the pull request! 🤝 ✒️ Just a reminder that the CLA has not yet been signed, and we'll need it before merging. Please [sign the CLA](https://www.influxdata.com/legal/cla/) when you get a chance, then post a comment here saying `!signed-cla`

!signed-cla

File: plugins/outputs/inlong/sample.conf [hunk 1]
```diff
@@ -0,0 +1,32 @@
+# Send telegraf metrics to Inlong
+[[outputs.inlong]]
+  ## From the inlong system, data Streams Group, it contains multiple data streams, and one Group represents
+  ## one data business unit.
+  inlong_group_id = "test_group"
```
Comment (DStrand1) [inline plugins/outputs/inlong/sample.conf:5]: Could you comment out any config options here that aren't required to provide?

File: plugins/outputs/inlong/inlong.go [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+				producer, err := dataproxy.NewClient(
+					dataproxy.WithGroupID(id),
+					dataproxy.WithURL(url),
+				)
+				if err != nil {
+					fmt.Println(err)
+					return nil, err
+				}
+				return producer, nil
+			},
```
Comment (DStrand1) [inline plugins/outputs/inlong/inlong.go:82]: Could you extract this function into a separate function for better readability?

**Review by DStrand1 — COMMENTED** [review]
Thanks for the PR! I have some initial comments if you could take a look

File: plugins/outputs/inlong/inlong_test.go [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+
+func NewMockProducer(groupID, managerURL string) (dataproxy.Client, error) {
+	p := &MockProducer{}
+	p.groupID = groupID
+	p.managerURL = managerURL
+	return p, nil
+}
+
+func TestInlong_Connect(t *testing.T) {
+	t.Run("", func(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (justinwwhuang) [reply plugins/outputs/inlong/inlong_test.go:43]: > Is this wrapping necessary in these tests?

No need, I'll simplify it

File: plugins/outputs/inlong/inlong.go (hunk 2 above)
Comment (justinwwhuang) [reply plugins/outputs/inlong/inlong.go:84]: > Could you extract this function into a separate function for better readability?

ok

File: plugins/outputs/inlong/sample.conf (hunk 1 above)
Comment (justinwwhuang) [reply plugins/outputs/inlong/sample.conf:5]: > Could you comment out any config options here that aren't required to provide?

ok

Comment (justinwwhuang) [reply plugins/outputs/inlong/sample.conf:5]: > Could you comment out any config options here that aren't required to provide?

Well, I found that retry_interval_ms is useless, I will delete it, but other configurations are necessary

File: plugins/outputs/inlong/README.md [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+See the [CONFIGURATION.md][CONFIGURATION.md] for more details.
+
+[CONFIGURATION.md]: ../../../docs/CONFIGURATION.md#plugins
+
+## Configuration
+
+```toml @sample.conf
+# Send telegraf metrics to Inlong
+[[outputs.inlong]]
+  ## From the inlong system, data Streams Group, it contains multiple data streams, and one Group represents
```
Comment (dockerzhang) [inline plugins/outputs/inlong/README.md:19]: inlong -> InLong

Comment (justinwwhuang) [reply plugins/outputs/inlong/README.md]: > inlong -> InLong

ok

File: plugins/outputs/inlong/README.md [hunk 5]
```diff
@@ -0,0 +1,44 @@
+# Inlong Output Plugin
+
+This plugin writes telegraf metrics to Inlong
```
Comment (dockerzhang) [inline plugins/outputs/inlong/README.md:3]:
```suggestion
This plugin writes telegraf metrics to [Apache InLong](https://inlong.apache.org/docs/next/introduction).
```

**Comment by justinwwhuang** [comment]
@DStrand1 Hello, I have made corresponding modifications to your comments. Please take a look again. Thank you

File: plugins/outputs/inlong/README.md [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+  stream_id = "test_stream"
+
+  ## The URL used to obtain the Inlong DataProxy IP list to which the data will be sent
+  manager_url = "http://127.0.0.1:8083"
+
+  ## Data format to output.
+  ## Each data format has its own unique set of configuration options, read
+  ## more about them here:
+  ## https://github.com/influxdata/telegraf/blob/master/docs/DATA_FORMATS_OUTPUT.md
+  ## Suggest using CSV format here, as Inlong is also processed in CSV format
```
Comment (DStrand1) [inline plugins/outputs/inlong/README.md:34]:
```suggestion
```

File: plugins/outputs/inlong/inlong.go [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+
+	"github.com/influxdata/telegraf"
+	"github.com/influxdata/telegraf/internal"
+	"github.com/influxdata/telegraf/plugins/outputs"
+)
+
+//go:embed sample.conf
+var sampleConfig string
+
+const ManagerURLSuffix = "/inlong/manager/openapi/dataproxy/getIpList"
```
Comment (DStrand1) [inline plugins/outputs/inlong/inlong.go:18]: This should be unexported

File: plugins/outputs/inlong/inlong.go [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+
+func init() {
+	outputs.Add("inlong", func() telegraf.Output {
+		return &Inlong{
+			producerFunc: NewProducer,
+		}
+	})
+}
+
+func NewProducer(groupID, managerURL string) (dataproxy.Client, error) {
```
Comment (DStrand1) [inline plugins/outputs/inlong/inlong.go:79]: This should be unexported

File: plugins/outputs/inlong/inlong.go (hunk 7 above)
Comment (justinwwhuang) [reply plugins/outputs/inlong/inlong.go:18]: get

File: plugins/outputs/inlong/inlong.go (hunk 8 above)
Comment (justinwwhuang) [reply plugins/outputs/inlong/inlong.go:79]: get

**Comment by justinwwhuang** [comment]
@DStrand1 Hello, I have made corresponding modifications to your comments. Please take a look again. Thank you

**Comment by justinwwhuang** [comment]
@DStrand1 I have made corresponding modifications to your comments. Please take a look again. Thank you

**Comment by justinwwhuang** [comment]
@DStrand1 Is there anything blocking this PR? I'd appreciate feedback to improve it.

File: plugins/outputs/inlong/README.md [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+
+```toml @sample.conf
+# Send telegraf metrics to Inlong
+[[outputs.inlong]]
+  ## From the Inlong system, data streams group, it contains multiple data streams, and one Group represents
+  ## one data business unit.
+  group_id = "test_group"
+
+  ## From the Inlong system, data stream, a stream has a specific data source, data format and data sink.
+  stream_id = "test_stream"
```
Comment (DStrand1) [inline plugins/outputs/inlong/README.md:25]: These two configuration option descriptions are a bit unclear to me, could you reword them?

File: plugins/outputs/inlong/inlong.go [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+	return nil
+}
+
+func init() {
+	outputs.Add("inlong", func() telegraf.Output {
+		return &Inlong{
+			producerFunc: newProducer,
+		}
+	})
+}
```
Comment (DStrand1) [inline plugins/outputs/inlong/inlong.go:77]: Nitpick: could you move this function to the very end of the file? This is more typical for our plugin file layout

**Review by DStrand1 — COMMENTED** [review]
Sorry for the slow response! Overall this looks good, I just have a couple remaining smaller comments

File: plugins/outputs/inlong/README.md (hunk 9 above)
Comment (justinwwhuang) [reply plugins/outputs/inlong/README.md]: > These two configuration option descriptions are a bit unclear to me, could you reword them?

I have completed the modification, please check if it meets your requirements

File: plugins/outputs/inlong/inlong.go (hunk 10 above)
Comment (justinwwhuang) [reply plugins/outputs/inlong/inlong.go]: > Nitpick: could you move this function to the very end of the file? This is more typical for our plugin file layout

done

File: plugins/outputs/inlong/inlong.go [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+		if err != nil {
+			return fmt.Errorf("could not serialize metric: %w", err)
+		}
+		err = i.producer.Send(context.Background(), dataproxy.Message{
+			GroupID:  i.GroupID,
+			StreamID: i.StreamID,
+			Payload:  b,
+		})
+		if err != nil {
+			return err
```
[the remark this answers is not in the document]
Comment (justinwwhuang) [reply plugins/outputs/inlong/inlong.go]: > Could you wrap this error with some more description?

I have completed the modification, please check if it meets your requirements

**Comment by justinwwhuang** [comment]
> Sorry for the slow response! Overall this looks good, I just have a couple remaining smaller comments
Thank you for your comment, I have made the necessary modifications according to it. The current test-go-mac failed to run, I don't quite understand what the error is.

**Comment by justinwwhuang** [comment]
@DStrand1 @srebhan Hello, can you help merge this PR? This PR has been in existence for a long time and has already resolved several code conflicts. If there are any issues, I am willing to make changes.

File: plugins/outputs/inlong/README.md [hunk 12]
```diff
@@ -0,0 +1,55 @@
+# Inlong Output Plugin
+
+This plugin writes telegraf metrics to
+[Apache InLong](https://inlong.apache.org/docs/next/introduction).
+
```
Comment (mstrandboge) [inline plugins/outputs/inlong/README.md:5]: In this top section, could you include some metadata tags for this plugin?
For example:
```md
⭐ Telegraf v1.35.0
🏷️ <tag>
💻 all
```
For the `<tag>` option, pick whichever apply (one or more):
`applications, cloud, containers, datastore, hardware, iot, logging, messaging, network, server, system, testing, web`
You can refer to other plugins to see some examples of this if you need

**Review by mstrandboge — COMMENTED** [review]
@justinwwhuang Thank you for bearing with us! I have just one comment left on the readme. On top of that, could you rebase this PR to resolve the merge conflict issue with the go.mod/go.sum and ensure the latest linters pass? Thanks!

File: plugins/outputs/inlong/README.md (hunk 12 above)
Comment (justinwwhuang) [reply plugins/outputs/inlong/README.md:5]: @mstrandboge hi, I have include metadata tags for this plugin as comment

File: go.mod [hunk 13]
```diff
@@ -245,6 +246,17 @@ require (
...
+require (
+	github.com/bwmarrin/snowflake v0.3.0 // indirect
+	github.com/go-resty/resty/v2 v2.16.2 // indirect
+	github.com/oxtoacart/bpool v0.0.0-20190530202638-03653db5a59c // indirect
+	github.com/panjf2000/ants/v2 v2.10.0 // indirect
+	github.com/panjf2000/gnet/v2 v2.6.3 // indirect
+	github.com/valyala/bytebufferpool v1.0.0 // indirect
+	github.com/zentures/cityhash v0.0.0-20131128155616-cdd6a94144ab // indirect
+	gopkg.in/natefinch/lumberjack.v2 v2.2.1 // indirect
+)
```
Comment (mstrandboge) [inline go.mod:258]: Can you merge this separate indirect block into the main indirect block?

**Review by mstrandboge — APPROVED** [review]
Just one small comment left, otherwise approved! Thanks for being patient and wrapping up this PR

File: go.mod (hunk 13 above)
Comment (justinwwhuang) [reply go.mod]: > Can you merge this separate indirect block into the main indirect block?

Hi, I have made the necessary changes according to the comments

**Review by dockerzhang — APPROVED** [review]

File: plugins/outputs/inlong/inlong_test.go [hunk 14]
```diff
@@ -0,0 +1,68 @@
...
+		},
+		map[string]interface{}{
+			"value": 42.0,
+		},
+		time.Unix(0, 0),
+	)
+	var metrics []telegraf.Metric
+	metrics = append(metrics, m)
+	require.NoError(t, i.Write(metrics))
+}
```
Comment (srebhan) [inline plugins/outputs/inlong/inlong_test.go:68]: To be honest, I don't know if this test adds any meaningful guard at all. The mock producer will not validate _any_ of the data sent nor any of the parameters. It would succeed even if you don't specify any URL or group or stream (as you do here) and without a connect! So you could do anything and this test would _always_ succeed! That's not a test in my view.

File: plugins/outputs/inlong/inlong.go [hunk 15]
```diff
@@ -0,0 +1,88 @@
...
+
+const managerURLSuffix = "/inlong/manager/openapi/dataproxy/getIpList"
+
+type Inlong struct {
+	GroupID    string          `toml:"group_id"`
+	StreamID   string          `toml:"stream_id"`
+	ManagerURL string          `toml:"manager_url"`
+	Log        telegraf.Logger `toml:"-"`
+
+	producerFunc func(groupId string, managerUrl string) (dataproxy.Client, error)
```
Comment (srebhan) [inline plugins/outputs/inlong/inlong.go:26]: This indirection through `producerFunc` seems to be introduced for testing. However, as it stands now, the tests add no real value so I would remove the indirection in favor of simplifying the code.

File: plugins/outputs/inlong/inlong.go [hunk 16]
```diff
@@ -0,0 +1,88 @@
...
+	i.producer = producer
+	return nil
+}
+
+func (i *Inlong) Close() error {
+	i.producer.Close()
+	return nil
+}
+
+func (i *Inlong) Write(metrics []telegraf.Metric) error {
```
Comment (srebhan) [inline plugins/outputs/inlong/inlong.go:53]: What happens if the Inlong instance goes down for whatever reason and comes up again e.g. due to network issues or service outages? Would the plugin reconnect automatically in this case?

**Review by srebhan — CHANGES_REQUESTED** [review]
Thank you very much @justinwwhuang for your contribution! Much appreciated. However, I do have some comments regarding the code. The major points are

1. There is no form of user-specified option checking e.g. if no or a malformed URL is specified, missing group- or stream-IDs...
2. The tests don't add any value as they will _always_ succeed even when the plugin is set-up with invalid configuration options.
3. I don't see any way to authenticate to the service. I can't imagine that _anyone_ will use Inlong in an production environment without authentication...

While I'm okay with adding authentication later (point 3), the former ones need some more work.

For option checking (point 1) I suggest to introduce an [`Init()` function](https://github.com/influxdata/telegraf/blob/d99b033395a379d88a5fd006eebe2fe38ba420a5/plugin.go#L16) to the plugin and check the values there with at least a test for

- provided URL is not empty
- provided URL is a valid http or https URL
- provided group-id is not empty
- provided stream-id is not empty

If the latter two are "optional" you can skip those or add a sensible default value (e.g. `"telegraf"`).

For testing (point 2) I do see three options actually:

1. Produce a mock implementation. This would require a mock of the manager "server" endpoint for getting the node IDs and a mock of the data-proxy send protocol. Information for producing the mock can be found in the inlong SDK I think together with some reverse engineering by running queries against a real instance and observe the responses. Certainly the hardest of the options but also the most rewarding one...
2. Setup an integration test using docker containers by "manually" composing the instance mimicking e.g. this [docker compose setup](https://github.com/apache/inlong/blob/master/docker/docker-compose/docker-compose.yml). This would guarantee that the plugin would _actually_ work with a _real_ Inlong instance...
3. If none of the above is possible, at least add some tests for invalid user options and skip all the always-succeeding mock-producer tests.

My preference would be a mix of 1 and 2 but that might be some effort.

So long story short, I would need at least the user-setting checks and a rudimentary test being implemented to be able to merge this PR.

File: plugins/outputs/inlong/inlong.go [hunk 17]
```diff
@@ -0,0 +1,88 @@
...
+
+func (i *Inlong) SetSerializer(serializer telegraf.Serializer) {
+	i.serializer = serializer
+}
+
+func (i *Inlong) Connect() error {
+	producer, err := i.producerFunc(i.GroupID, i.ManagerURL+managerURLSuffix)
+	if err != nil {
+		return &internal.StartupError{Err: err, Retry: true}
+	}
```
[the remark this answers is not in the document]
Comment (justinwwhuang) [reply plugins/outputs/inlong/inlong.go:43]: Sorry, i will add an Init() function to exclude all "invalid" configs ahead of time

File: plugins/outputs/inlong/inlong.go [hunk 18]
```diff
@@ -0,0 +1,88 @@
...
+}
+
+func (*Inlong) SampleConfig() string {
+	return sampleConfig
+}
+
+func (i *Inlong) SetSerializer(serializer telegraf.Serializer) {
+	i.serializer = serializer
+}
+
```
[the remark this answers is not in the document]
Comment (justinwwhuang) [reply plugins/outputs/inlong/inlong.go:38]: ok

**Comment by srebhan** [comment]
@justinwwhuang please re-request a review for me to get a notification and guarantee fast update-review cycles... You are already waiting for too long... :-(

File: plugins/outputs/inlong/inlong.go (hunk 16 above)
Comment (justinwwhuang) [reply plugins/outputs/inlong/inlong.go:73]: Yes, SDK has the ability to automatically reconnect

File: plugins/outputs/inlong/inlong_test.go [hunk 19]
```diff
@@ -0,0 +1,68 @@
...
+func NewMockProducer(groupID, managerURL string) (dataproxy.Client, error) {
+	p := &MockProducer{}
+	p.groupID = groupID
+	p.managerURL = managerURL
+	return p, nil
+}
+
+func TestInlong_Connect(t *testing.T) {
+	i := &Inlong{producerFunc: NewMockProducer}
+	require.NoError(t, i.Connect())
```
[the remark this answers is not in the document]
Comment (justinwwhuang) [reply plugins/outputs/inlong/inlong_test.go]: ok, I  will fix it

File: plugins/outputs/inlong/inlong.go (hunk 15 above)
Comment (justinwwhuang) [reply plugins/outputs/inlong/inlong.go:29]: Well, I will use it at TestInlong_Connect

File: plugins/outputs/inlong/inlong_test.go (hunk 14 above)
Comment (justinwwhuang) [reply plugins/outputs/inlong/inlong_test.go]: ok, I will add specific data verification

File: plugins/outputs/inlong/inlong.go [hunk 20]
```diff
@@ -0,0 +1,111 @@
...
+		return errors.New("managerURL must not be empty")
+	}
+	u, err := url.Parse(i.ManagerURL)
+	if err != nil || u.Scheme == "" || u.Host == "" {
+		return fmt.Errorf("managerURL %v invalid", i.ManagerURL)
+	}
+	if !strings.HasPrefix(u.Scheme, "http") {
+		return fmt.Errorf("managerURL %v should start with http", i.ManagerURL)
+	}
+	return nil
```
Comment (srebhan) [inline plugins/outputs/inlong/inlong.go:55]: How about
```suggestion
	u, err := url.Parse(i.ManagerURL)
	if err != nil {
		return fmt.Errorf("invalid URL %q", i.ManagerURL)
	}
	switch u.Scheme {
	case "http", "https":
		if u.Host == "" {
			return fmt.Errorf("no host in URL %q", i.ManagerURL)
		}
	default:
		return fmt.Errorf("invalid schema in URL %q", i.ManagerURL)
	}	
	u.ManagerURL = u.String(u.JoinPath("inlong", "manager", "openapi", "dataproxy", "getIpList"))
	
	return nil
```

File: plugins/outputs/inlong/inlong.go [hunk 21]
```diff
@@ -0,0 +1,111 @@
...
+	if err != nil {
+		return nil, fmt.Errorf("creating a producer for group %q at %q failed: %w", groupID, managerURL, err)
+	}
+	return producer, nil
+}
+
+func init() {
+	outputs.Add("inlong", func() telegraf.Output {
+		return &Inlong{
+			producerFunc: newProducer,
```
Comment (srebhan) [inline plugins/outputs/inlong/inlong.go:108]: Please remove the `producerFunc` indirection and directly create the new client. I will take care of the tests once this is merged...

File: plugins/outputs/inlong/inlong_test.go [hunk 22]
```diff
@@ -0,0 +1,94 @@
...
+}
+
+func getSeparator() string {
+	switch runtime.GOOS {
+	case "windows":
+		return "\r\n"
+	default:
+		return "\n"
+	}
+}
```
Comment (srebhan) [inline plugins/outputs/inlong/inlong_test.go:94]: I don't think this whole stuff is worth the effort. Again you test nothing besides that the serializer is working and that what you put in is coming out because you bypass all inlong-specific code. How about some rudimentary testing and we add more sophisticated tests later? Like

```suggestion
package inlong

import (
	"context"
	"runtime"
	"strings"
	"testing"
	"time"

	"github.com/stretchr/testify/require"

	"github.com/influxdata/telegraf"
)

func TestInvalidParameters(t *testing.T) {
	tests := []struct{
		name     string
		url      string
		gid      string
		sid      string
		expected string
	}{
		{
			name:     "all empty",
			expected: "'url' must not be empty",
		},
		{
			name:     "invalid url scheme",
			url:      "unix://localhost",
			expected: "invalid schema in URL",
		},
		{
			name:     "no host",
			url:      "http://:443",
			expected: "no host in URL",
		},
		{
			name:     "group id empty",
			url:      "http://localhost",
			expected: "'group_id' must not be empty",
		},	
		{
			name:     "stream id empty",
			url:      "http://localhost",
			gid:      "test",
			expected: "'stream_id' must not be empty",
		},	
	}
	
        for _, tt := range tests {
                t.Run(tt.name, func(t *testing.T) {
			plugin := &Inlong{
				ManagerURL: tt.url,
				GroupID:    tt.gid,
				StreamID:   tt.sid,
			},
			require.ErrorContains(t, plugin.Init(), tt.expected)
                })
        }
}

func TestValidURLs(t *testing.T) {
	tests := []struct{
		name     string
		url      string
	}{
		{
			name: "http url scheme",
			url:  "http://localhost",
		},
		{
			name: "http url scheme with port",
			url:  "http://localhost:8080",
		},
		{
			name: "http url scheme with port and path",
			url:  "http://localhost:8080/foo",
		},
		{
			name: "https url scheme",
			url:  "https://localhost",
		},
		{
			name: "https url scheme with port",
			url:  "https://localhost:8080",
		},
		{
			name: "https url scheme with port and path",
			url:  "https://localhost:8080/foo",
		},
	}
	
        for _, tt := range tests {
                t.Run(tt.name, func(t *testing.T) {
			plugin := &Inlong{
				ManagerURL: tt.url,
				GroupID:    "test",
				StreamID:   "test",
			},
			require.NoError(t, plugin.Init())
                })
        }
}

**Review by srebhan — COMMENTED** [review]
Thanks for the update @justinwwhuang! Some more suggestion and requests...

File: plugins/outputs/inlong/inlong.go (hunk 21 above)
Comment (justinwwhuang) [reply plugins/outputs/inlong/inlong.go:115]: ok

**Comment by srebhan** [comment]
@justinwwhuang as I wrote last time: **PLEASE DO NOT OPEN ISSUES TO NOTIFY ANYONE!!!!** You are cluttering our issues and make it even harder to find _anything_ in there! Please use the `Re-request Review` button in the PR! I will get a notification if you do so!

**Comment by justinwwhuang** [comment]
> @justinwwhuang as I wrote last time: **PLEASE DO NOT OPEN ISSUES TO NOTIFY ANYONE!!!!** You are cluttering our issues and make it even harder to find _anything_ in there! Please use the `Re-request Review` button in the PR! I will get a notification if you do so!

Sorry, I made a mistake in my operation

File: plugins/outputs/inlong/inlong.go [hunk 23]
```diff
@@ -0,0 +1,106 @@
...
+		return errors.New("'stream_id' must not be empty")
+	}
+	parsedURL, err := url.Parse(i.ManagerURL)
+	if err != nil {
+		return fmt.Errorf("invalid URL: %s", i.ManagerURL)
+	}
+
+	switch parsedURL.Scheme {
+	case "http", "https":
+		fmt.Println(parsedURL.Hostname())
```
[the remark this answers is not in the document]
Comment (justinwwhuang) [reply plugins/outputs/inlong/inlong.go:51]: Sorry, it's true

File: plugins/outputs/inlong/inlong.go [hunk 24]
```diff
@@ -0,0 +1,106 @@
...
+}
+
+func (i *Inlong) Connect() error {
+	producer, err := dataproxy.NewClient(
+		dataproxy.WithGroupID(i.GroupID),
+		dataproxy.WithURL(i.ManagerURL),
+	)
+	if err != nil {
+		return &internal.StartupError{Err: err, Retry: true}
+	}
```
[the remark this answers is not in the document]
Comment (justinwwhuang) [reply plugins/outputs/inlong/inlong.go:74]: > Are you **_really_** sure _all_ errors returned by `dataproxy.NewClient` are retryable?!?!?

In addition to basic configuration errors, failures usually occur only when the inlong system is under high load, and are usually attempted again.

[the remark this answers is not in the document]
Comment (justinwwhuang) [reply plugins/outputs/inlong/inlong.go:74]: > > Are you **_really_** sure _all_ errors returned by `dataproxy.NewClient` are retryable?!?!?
> 
> In addition to basic configuration errors, failures usually occur only when the inlong system is under high load, and are usually attempted again.

Perhaps some error logs should be output here, but the data should not be directly discarded

**Comment by justinwwhuang** [comment]
@srebhan hi, could you help to rerun this pr ? it seems failed due to lack of private ip, I have no right to rerun it

File: plugins/outputs/inlong/inlong.go (hunk 24 above)
[the remark this answers is not in the document]
Comment (justinwwhuang) [reply plugins/outputs/inlong/inlong.go:74]: > Are you **_really_** sure _all_ errors returned by `dataproxy.NewClient` are retryable?!?!?

In fact, when we encounter initialization failures in practical use, we usually perform timed retries and output error logs or alarms

File: plugins/outputs/inlong/sample.conf [hunk 25]
```diff
@@ -0,0 +1,27 @@
...
+  data_format = "csv"
+
+  ## The delimiter used when serializing data in CSV format needs to be consistent with the delimiter
+  ## configured for Inlong, so that the data can be parsed properly after it reaches Inlong.
+  ## It can be a space, vertical bar (|), comma (,), semicolon (;), asterisk (*), double quotes ("), etc.
+  csv_separator = "|"
+
+  ## The final output field order here needs to be consistent with the field order defined by the data
+  ## stream in Inlong
+  csv_columns = ["field.key","file.value"]
```
Comment (srebhan) [inline plugins/outputs/inlong/sample.conf:27]: Before running `make docs` you should have copied the README config to here to not revert all changes. ;-) You need
```suggestion
# Send telegraf metrics to Apache Inlong
[[outputs.inlong]]
  ## Manager URL to obtain the Inlong data-proxy IP list for sending the data
  url = "http://127.0.0.1:8083"

    ## Unique identifier for the data-stream group
  group_id = "telegraf"  

  ## Unique identifier for the data stream within its group
  stream_id = "telegraf"  

  ## Data format to output.
  ## Each data format has its own unique set of configuration options, read
  ## more about them here:
  ## https://github.com/influxdata/telegraf/blob/master/docs/DATA_FORMATS_OUTPUT.md
  # data_format = "influx"
```

this in here and afterwards run `make docs` to update the `README.md` file.

File: plugins/outputs/inlong/inlong.go [hunk 26]
```diff
@@ -0,0 +1,107 @@
...
+}
+
+func (i *Inlong) Connect() error {
+	producer, err := dataproxy.NewClient(
+		dataproxy.WithGroupID(i.GroupID),
+		dataproxy.WithURL(i.ManagerURL),
+	)
+	if err != nil {
+		i.Log.Errorf("Connect to manager %v group_id %v error %v", i.ManagerURL, i.GroupID, err)
+		return &internal.StartupError{Err: err, Retry: true}
```
Comment (srebhan) [inline plugins/outputs/inlong/inlong.go:74]: Please do not log the error, it will be logged automatically when returning it so better do
```suggestion
		return &internal.StartupError{
			Err: fmt.Errorf("connecting to manager %q with group-ID %q failed: %w", i.ManagerURL, i.GroupID, err)
			Retry: true,
		}
```

File: plugins/outputs/inlong/inlong.go [hunk 27]
```diff
@@ -0,0 +1,107 @@
...
+		}
+		err = i.producer.Send(context.Background(), dataproxy.Message{
+			GroupID:  i.GroupID,
+			StreamID: i.StreamID,
+			Payload:  b,
+		})
+		if err != nil {
+			return fmt.Errorf("could not send metric to GroupID %s StreamID %s: %w", i.GroupID, i.StreamID, err)
+		}
+	}
```
Comment (srebhan) [inline plugins/outputs/inlong/inlong.go:99]: Do we really need to send this metric-by-metric? Can't we send all metrics together?

**Review by srebhan — COMMENTED** [review]
Thanks for the update @justinwwhuang! Unfortunately, you were overwriting the config changes by `make docs`. I restored it in the `sample.conf` file. Please _only_ edit the `sample.conf` file when modifying the config and run `make docs` afterwards.

Additionally, I have one suggestion and one more question... We are very close for this to be mergable...

File: plugins/outputs/inlong/inlong.go (hunk 27 above)
Comment (justinwwhuang) [reply plugins/outputs/inlong/inlong.go:101]: > Do we really need to send this metric-by-metric? Can't we send all metrics together?

The SDK will assemble and send in batches internally

Comment (srebhan) [reply plugins/outputs/inlong/inlong.go:101]: Yeah what I meant is, you are currently serializing each metric individually, but you also could serialize the whole set of metrics and send it with one call like
```suggestion
	b, err := i.serializer.SerializeBatch(metrics)
	if err != nil {
		return fmt.Errorf("could not serialize metrics: %w", err)
	}
	err = i.producer.Send(context.Background(), dataproxy.Message{
		GroupID:  i.GroupID,
		StreamID: i.StreamID,
		Payload:  b,
	})
	if err != nil {
		return fmt.Errorf("could not send metric to GroupID %s StreamID %s: %w", i.GroupID, i.StreamID, err)
	}
```

Comment (justinwwhuang) [reply plugins/outputs/inlong/inlong.go:101]: > Yeah what I meant is, you are currently serializing each metric individually, but you also could serialize the whole set of metrics and send it with one call like
> 
> I'm not sure if that make sense to consumers of the message, that's why I'm asking. If both variants make sense, we can implement it later in a separate PR.

This will cause the inlong platform to recognize these multiple messages as one, and I mistakenly merged them into it. I will revert again

**Review by srebhan — APPROVED** [review]
Thanks for your contribution @justinwwhuang! I'll fix the two small formatting issue so we can continue early.

**Comment by srebhan** [comment]
@mstrandboge could you please give this another quick review as the code changed significantly and then merge it!?

**Review by mstrandboge — APPROVED** [review]

**Comment by mstrandboge** [comment]
@justinwwhuang One last thing, could you resolve the merge conflicts with go.mod and go.sum? Thank you!

**Comment by justinwwhuang** [comment]
> @justinwwhuang One last thing, could you resolve the merge conflicts with go.mod and go.sum? Thank you!

Hello, I have already resolved it. Please check
